lib/cyrusdb_twom.c

```diff
@@ -0,0 +1,2596 @@
+    uint32_t csum = file->csum(ptr, HEADLEN(ptr));
+    if (csum != HEADCSUM(ptr)) {
+        xsyslog(LOG_ERR, "DBERROR: invalid head checksum",
+                         "filename=<%s> offset=<%08llX>",
+                         db->fname, (LLU)offset);
+        return CYRUSDB_IOERROR;
+    }
+
+    return 0;
+}
+
+#ifdef HAVE_DECLARE_OPTIMIZE
+static inline void _setloc(struct tm_file *file, char *ptr, uint8_t level, size_t offset)
+    __attribute__((optimize("-O3")));
+#endif
+static inline void _setloc(struct tm_file *file, char *ptr, uint8_t level, size_t offset)
+{
+    char *addr = ptr + ptroffset[TYPE(ptr)];
+
+    if (level) {
+        // positions past the start
+        addr += 8 * (level + 1);
+    }
+    else {
+        /* level zero is special */
+        size_t val0 = NEXT0(ptr, 0);
+        size_t val1 = NEXT0(ptr, 1);
+
+        size_t end = file->header.current_size;
+        /* already this transaction, update this one */
+        if (val0 < end && (val1 >= end || val0 > val1))
+            addr += 8; // conditions to write to val1
+    }
+
+    *((uint64_t *)(addr)) = htonll(offset);
+}
+
+#ifdef HAVE_DECLARE_OPTIMIZE
+static inline void _recsum(struct tm_file *file, char *ptr)
+    __attribute__((optimize("-O3")));
+#endif
+static inline void _recsum(struct tm_file *file, char *ptr)
+{
+    size_t headlen = HEADLEN(ptr);
+    uint32_t newcsum = file->csum(ptr, headlen);
+    *((uint32_t *)(ptr + headlen)) = htonl(newcsum);
+}
+
+/* finds a record, either an exact match or the record
+ * immediately before */
+#ifdef HAVE_DECLARE_OPTIMIZE
+static int locate(struct dbengine *db, struct tm_loc *loc, const char *key, size_t keylen)
+    __attribute__((optimize("-O3")));
+#endif
+static int locate(struct dbengine *db, struct tm_loc *loc, const char *key, size_t keylen)
+{
+    size_t offset = 0;
+    uint8_t level = MAXLEVEL;
+    int cmp = -1; /* never found a thing! */
+    const char *ptr = NULL;
+    struct tm_file *file = loc->file;
+
+    // refcount the map where this location is placed
+    loc->offset = DUMMY_OFFSET;
+    loc->backloc[level] = DUMMY_OFFSET;
+    loc->is_exactmatch = 0;
+
+    /* start with the dummy */
+    const char *locptr = safeptr(loc, DUMMY_OFFSET);
+    if (!locptr) return CYRUSDB_IOERROR; // invalid ptr
+
+    /* special case start pointer for efficiency */
+    if (!keylen) {
+        while (level) {
+            loc->backloc[level-1] = DUMMY_OFFSET;
+            level--;
+        }
+        return 0;
+    }
+
+    while (level) {
+        size_t next = ADVANCE(loc, locptr, level - 1);
+
+        loc->backloc[level-1] = loc->offset;
+
+        if (next != offset) {
+            offset = next;
+            if (offset) {
+                ptr = safeptr(loc, offset);
+                if (!ptr) return CYRUSDB_IOERROR;
+                assert(LEVEL(ptr) >= level);
+
+                cmp = file->compar(KEYPTR(ptr), KEYLEN(ptr),
+                                   key, keylen);
+
```

Comment:
I'm unclear what this function is doing so excuse my ignorance: If cmp == 0, doesn't that mean we've found an exact match and can bail early? It seems like instead we keep going...